sycl/include/sycl/ext/intel/experimental/esimd/simd.hpp

```diff
@@ -45,25 +45,20 @@ template <typename Ty, int N> class simd {
   /// The number of elements in this simd object.
   static constexpr int length = N;
 
-  // TODO @rolandschulz
-  // Provide examples why constexpr is needed here.
-  //
   /// @{
   /// Constructors.
-  constexpr simd() = default;
-  template <typename SrcTy> constexpr simd(const simd<SrcTy, N> &other) {
-    if constexpr (std::is_same<SrcTy, Ty>::value)
-      set(other.data());
-    else
-      set(__builtin_convertvector(other.data(), detail::vector_type_t<Ty, N>));
+  simd() = default;
+  simd(const simd &other) { set(other.data()); }
```

Comment:
@rolandschulz suggested to make it default too, which makes sense unless there are unforeseen complications like bad code gen. Are there any? If yes, please add a comment. Otherwise, =default should work.

Reply:
Yes, but `set` is calling `__esimd_vstore` for device compilation. Let me double-check.

Reply:
@rolandschulz, @DenisBakhvalov - this patch luckily fixes a real problem. `constexpr` leads to creating a constant global of simd wrapper type, e.g:
```
@vc = private unnamed_addr addrspace(1) constant %"esimd::simd" \{ { <16 x i16> <i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1> }, align 32
...
call void @llvm.memcpy.p0i8.p0i8.i64(i8* %0, i8* bitcast (%class.simd* @vc to i8*), i64 32, i1 false)
```
which is not handled properly by the BE. W/o {{constexpr}}, the constant becomes an immediate of the primitive vector type, and memcpy is replaced with store. And this is handled by the BE correctly.

Reply:
I checked on a simplified sample - `simd(const simd &other) = default` will lead to a memcpy compared to `__esimd_vstore` (as Denis mentions above) done today, which may cause CG problems. Since keeping it non-default does not change compatibility, we can add =default later if we find a proper way. @rolandschulz, I'm approving and merging, if you still have concerns - we can discuss and fix as a separate PR.

Reply:
It seems like a good temporary solution. Note that it isn't trivial if is isn't `=default`. And by not being trivial it can't be used for those things limited to trivial types (e.g. bit_cast). In general it is also an ABI break to change whether a type is trivial because some ABIs CC is different for trivial vs non-trivial types. Not sure whether any of this is a concern here.